File: haskell/src/SwiftNav/SBP/Ndb.hs
{-# OPTIONS_GHC -fno-warn-unused-imports #-}
{-# LANGUAGE NoImplicitPrelude           #-}
{-# LANGUAGE TemplateHaskell             #-}
{-# LANGUAGE RecordWildCards             #-}

-- |
-- Module:      SwiftNav.SBP.Ndb
-- Copyright:   Copyright (C) 2015-2018 Swift Navigation, Inc.
-- License:     LGPL-3
-- Maintainer:  Swift Navigation <dev@swiftnav.com>
-- Stability:   experimental
-- Portability: portable
--
-- \<Messages for logging NDB events. \>

module SwiftNav.SBP.Ndb
  ( module SwiftNav.SBP.Ndb
  ) where

import BasicPrelude
import Control.Lens
import Control.Monad.Loops
import Data.Binary
import Data.Binary.Get
import Data.Binary.IEEE754
import Data.Binary.Put
import Data.ByteString.Lazy    hiding (ByteString)
import Data.Int
import Data.Word
import SwiftNav.SBP.TH
import SwiftNav.SBP.Types
import SwiftNav.SBP.Gnss

{-# ANN module ("HLint: ignore Use camelCase"::String) #-}
{-# ANN module ("HLint: ignore Redundant do"::String) #-}
{-# ANN module ("HLint: ignore Use newtype instead of data"::String) #-}


msgNdbEvent :: Word16
msgNdbEvent = 0x0400

-- | SBP class for message MSG_NDB_EVENT (0x0400).
--
-- This message is sent out when an object is stored into NDB. If needed
-- message could also be sent out when fetching an object from NDB.
data MsgNdbEvent = MsgNdbEvent
  { _msgNdbEvent_recv_time     :: !Word64
    -- ^ HW time in milliseconds.
  , _msgNdbEvent_event         :: !Word8
    -- ^ Event type.
  , _msgNdbEvent_object_type   :: !Word8
    -- ^ Event object type.
  , _msgNdbEvent_result        :: !Word8
    -- ^ Event result.
  , _msgNdbEvent_data_source   :: !Word8
    -- ^ Data source for STORE event, reserved for other events.
  , _msgNdbEvent_object_sid    :: !GnssSignal
    -- ^ GNSS signal identifier, If object_type is Ephemeris OR Almanac, sid
    -- indicates for which signal the object belongs to. Reserved in other
    -- cases.
  , _msgNdbEvent_src_sid       :: !GnssSignal
    -- ^ GNSS signal identifier, If object_type is Almanac, Almanac WN, Iono OR
    -- L2C capabilities AND data_source is NDB_DS_RECEIVER sid indicates from
    -- which SV data was decoded. Reserved in other cases.
  , _msgNdbEvent_original_sender :: !Word16
    -- ^ A unique identifier of the sending hardware. For v1.0, set to the 2
    -- least significant bytes of the device serial number, valid only if
    -- data_source is NDB_DS_SBP. Reserved in case of other data_source.
  } deriving ( Show, Read, Eq )

instance Binary MsgNdbEvent where
  get = do
    _msgNdbEvent_recv_time <- getWord64le
    _msgNdbEvent_event <- getWord8
    _msgNdbEvent_object_type <- getWord8
    _msgNdbEvent_result <- getWord8
    _msgNdbEvent_data_source <- getWord8
    _msgNdbEvent_object_sid <- get
    _msgNdbEvent_src_sid <- get
    _msgNdbEvent_original_sender <- getWord16le
    pure MsgNdbEvent {..}

  put MsgNdbEvent {..} = do
    putWord64le _msgNdbEvent_recv_time
    putWord8 _msgNdbEvent_event
    putWord8 _msgNdbEvent_object_type
    putWord8 _msgNdbEvent_result
    putWord8 _msgNdbEvent_data_source
    put _msgNdbEvent_object_sid
    put _msgNdbEvent_src_sid
    putWord16le _msgNdbEvent_original_sender

$(makeSBP 'msgNdbEvent ''MsgNdbEvent)
$(makeJSON "_msgNdbEvent_" ''MsgNdbEvent)
$(makeLenses ''MsgNdbEvent)

File: kaitai/nim/ndb.nim
import kaitai_struct_nim_runtime
import options
import gnss

type
  Ndb* = ref object of KaitaiStruct
    `parent`*: KaitaiStruct
  Ndb_MsgNdbEvent* = ref object of KaitaiStruct
    `recvTime`*: uint64
    `event`*: uint8
    `objectType`*: uint8
    `result`*: uint8
    `dataSource`*: uint8
    `objectSid`*: Gnss_GnssSignal
    `srcSid`*: Gnss_GnssSignal
    `originalSender`*: uint16
    `parent`*: Sbp_Message

proc read*(_: typedesc[Ndb], io: KaitaiStream, root: KaitaiStruct, parent: KaitaiStruct): Ndb
proc read*(_: typedesc[Ndb_MsgNdbEvent], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Ndb_MsgNdbEvent


proc read*(_: typedesc[Ndb], io: KaitaiStream, root: KaitaiStruct, parent: KaitaiStruct): Ndb =
  template this: untyped = result
  this = new(Ndb)
  let root = if root == nil: cast[Ndb](this) else: cast[Ndb](root)
  this.io = io
  this.root = root
  this.parent = parent


proc fromFile*(_: typedesc[Ndb], filename: string): Ndb =
  Ndb.read(newKaitaiFileStream(filename), nil, nil)


##[
This message is sent out when an object is stored into NDB. If needed
message could also be sent out when fetching an object from NDB.

]##
proc read*(_: typedesc[Ndb_MsgNdbEvent], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Ndb_MsgNdbEvent =
  template this: untyped = result
  this = new(Ndb_MsgNdbEvent)
  let root = if root == nil: cast[Ndb](this) else: cast[Ndb](root)
  this.io = io
  this.root = root
  this.parent = parent


  ##[
  HW time in milliseconds.

  ]##
  let recvTimeExpr = this.io.readU8le()
  this.recvTime = recvTimeExpr

  ##[
  Event type.

  ]##
  let eventExpr = this.io.readU1()
  this.event = eventExpr

  ##[
  Event object type.

  ]##
  let objectTypeExpr = this.io.readU1()
  this.objectType = objectTypeExpr

  ##[
  Event result.

  ]##
  let resultExpr = this.io.readU1()
  this.result = resultExpr

  ##[
  Data source for STORE event, reserved for other events.

  ]##
  let dataSourceExpr = this.io.readU1()
  this.dataSource = dataSourceExpr

  ##[
  GNSS signal identifier, If object_type is Ephemeris OR Almanac, sid
indicates for which signal the object belongs to. Reserved in other
cases.

  ]##
  let objectSidExpr = Gnss_GnssSignal.read(this.io, this.root, this)
  this.objectSid = objectSidExpr

  ##[
  GNSS signal identifier, If object_type is Almanac, Almanac WN, Iono
OR L2C capabilities AND data_source is NDB_DS_RECEIVER sid indicates
from which SV data was decoded. Reserved in other cases.

  ]##
  let srcSidExpr = Gnss_GnssSignal.read(this.io, this.root, this)
  this.srcSid = srcSidExpr

  ##[
  A unique identifier of the sending hardware. For v1.0, set to the 2
least significant bytes of the device serial number, valid only if
data_source is NDB_DS_SBP. Reserved in case of other data_source.

  ]##
  let originalSenderExpr = this.io.readU2le()
  this.originalSender = originalSenderExpr

proc fromFile*(_: typedesc[Ndb_MsgNdbEvent], filename: string): Ndb_MsgNdbEvent =
  Ndb_MsgNdbEvent.read(newKaitaiFileStream(filename), nil, nil)

